subiquity/server/controllers/snaplist.py

```diff
@@ -169,13 +197,22 @@ def make_autoinstall(self):
         return [attr.asdict(sel) for sel in self.model.selections]
 
     async def GET(self, wait: bool = False) -> SnapListResponse:
-        if self.loader.failed or not self.app.base_model.network.has_network:
+        if self.loader.fetch_list_failed() \
+                or not self.app.base_model.network.has_network:
             return SnapListResponse(status=SnapCheckState.FAILED)
-        if not self.loader.snap_list_fetched and not wait:
+        if not self.loader.fetch_list_completed() and not wait:
             return SnapListResponse(status=SnapCheckState.LOADING)
-        await self.loader.get_snap_list_task().wait()
-        if self.loader.failed or not self.app.base_model.network.has_network:
-            return SnapListResponse(status=SnapCheckState.FAILED)
+        # Let's wait for the task to be completed.
+        # If the loader gets restarted, the cancellation should be absorbed.
+        while True:
+            await self.loader.load_list_task_created.wait()
+            try:
+                await asyncio.shield(self.loader.get_snap_list_task())
+                break
```

Comment:
I slightly prefer to express this pattern by having the try: block end with "else: break" but that's very minor.

Reply:
Oh yeah, that is better!